man/zipnote.1
.TH zipnote 1 "v3.0 of 8 May 2008"
.SH NAME
zipnote \- write the comments in zipfile to stdout, edit comments and rename files in zipfile

.SH SYNOPSIS
.I zipnote
.RB [ \-w ]
.RB [ \-b\ path ]
.RB [ \-h ]
.RB [ \-v ]
.RB [ \-L ]
zipfile

.SH ARGUMENTS
zipfile  Zipfile to read comments from or edit.

.SH OPTIONS
.TP
.BI \-w
Write comments to a zipfile from stdin (see below).
.TP
.BI \-b\ \fRpath
Use path for the temporary zip file.
.TP
.BI \-h
Show a short help.
.TP
.BI \-v
Show version information.
.TP
.BI \-L
Show software license.

.SH DESCRIPTION
.I zipnote
writes the comments in a zipfile to stdout.  This is the default mode.  A second mode
allows updating the comments in a zipfile as well as allows changing the names
of the files in the zipfile.  These modes are described below.

.SH EXAMPLES
To write all comments in a zipfile to stdout use for example
.LP
.nf
     zipnote foo.zip > foo.tmp
.fi
.LP
This writes all comments in the zipfile
.I foo.zip
to the file
.I foo.tmp
in a specific format.

.LP
If desired, this file can then be edited to change the comments and then used
to update the zipfile.
.LP
.nf
     zipnote -w foo.zip < foo.tmp
.fi
.LP
The names of the files in the zipfile can also be changed in this way.  This is done by
following lines like
.nf
     "@ name"
.fi
in the created temporary file (called
.I foo.tmp
here) with lines like
.nf
     "@=newname"
.fi
and then using the -w option as above.

.SH BUGS
The temporary file format is rather specific and zipnote is rather picky about it.
It should be easier to change file names in a script.

Does not yet support large (> 2 GB) or split archives.

.SH SEE ALSO
zip(1), unzip(1)
.SH AUTHOR
Info-ZIP


CHANGES
------------------------- August 7th 1996 version 2.2a ------------------
 1. QDOS port (Jonathan Hudson)
 2. win32 volumelabel handling (Paul)
 3. VM/CMS clean up (Greg Hartwig)
 4. leading "../" in internal filenames are allowed (Paul)
 5. System V packages support (John Bush)
 6. Fix handling of atx in zipup() (Onno, Greg)
 7. Fixed typo that caused zip -R to dump core (Onno)
 8. msdos/makefile.dj2: fix for command line too long when linking zip.exe
 9. win95 long filename support with djgpp v2 (Onno, Kimio Itoh)
------------------------- August 9th 1996 version 2.2b ------------------
 1. windll: use wiz instead of wizip (Mike)
 2. use z->name NOT z->zname to open files (Onno, Mike)
------------------------ September 1st 1996 version 2.2c ------------------
 1. windll: use fprintf instead of putc to send data to std{out,err} (Mike)
 2. os2: make borlandc version detection equal to unzip 5.30d (Kai Uwe)
 3. use #elif constructions for msdos,os2 and win32 compiler detection (Onno)
 4. fix for incorrect free in zip.c (Onno, Mike, Steve)
 5. BeBox port from Chris
 6. unix/{configure,Makefile} fixes for SCO Xenix 286 (Tom Schmidt)
 7. remove zilog entry from unix/Makefile (Onno)
 8. man page fixes (Tom Schmidt)
 9. SCO ODT {3,5} fixes (Bill Davidsen)
------------------------ October 8th 1996 version 2.2d ------------------
 1. Fix bug in QDOS patch that broke zipsplit.c (Onno, Paul)
 2. Fix a couple of warnings from BorlandC (Mike)
 3. msdos/makefile.wat: Delete some more files when cleaning up (Paul)
 4. store msdos volumelabels without a dot in them (Paul)
 5. clean up of unix/{Makefile,configure,packaging} (Tom Schmidt)
 6. make QDOS port case independent (Jonathan Hudson)
 7. new amiga SASC makefile (Walter Haidinger)
 8. don't truncate filenames in win32's in2ex() (Paul)
 9. os2/makefile.os2 update for emx 0.9c (Kai Uwe)
10. password() function for QDOS (Jonathan)
11. fix the last(?) free() related bug (Mike)
12. win32: security descriptors operations (Scott Field)
13. win32: FILE_SHARE_DELETE is not defined in some win32 compilers (Onno)
14. win32: fix makefile.wat to include nt.c (Onno)
------------------------ January 17th 1997 version 2.2e ------------------
 1. define USE_CASE_MAP in osdep.h for those ports that need it (Onno)
 2. define PROCNAME in osdep.h for those ports that need it (Onno)
 3. wild() prototype decl only if PROCNAME defined => delete MSVMS define (Onno)
 4. add DOS EMX makefile (E-Yen Tan)
 5. include <qdos.h> a little earlier in qdos/qdos.c (Jonathan)
 6. add ttyio.o to OBJZ in qdos/Makefile.qdos (Jonathan)
 7. remove unused fprintebc define from zip.c (Onno)
 8. use the right password routine in ttyio.c for unzip (Mike)
 9. BeOS update from Chris
10. Fix for 'zip -r foo x:' (Paul)
11. Fix library bug on beos (Chris)
12. Fix calculating version number (kitoh_@mix.or.jp, Walter Haidinger)
13. IsWinNT always returned TRUE (Mike)
14. Windll update from Mike
15. Improved crc routines for x86 from Scott Field
16. Detect in unix/configure if we can use crc_i386.S (Onno)
17. Fix spurious internal logic error (Paul)
18. Fix to include directory names on the Acorn when needed (Sergio)
19. include zip.h in mvs.h (Onno, George Carr)
20. add workaround for AZTEC C compiler bug to revision.h (Paul, Walter)
21. MVS doesn't have rmdir (George Carr)
22. define and use USE_ZIPMAIN for WINDLL en VM_CMS (Onno)
23. Fixes from Greg Hartwig to make CMS standalone versions possible.
24. Move OS specific encryption stuff to the os specific directories (Christian)
25. Change password fetching interface in ttyio and crypt (Christian)
26. Update emx support for 0.9c (Christian)
27. Define WINDLL instead of MSWIN (Christian)
28. Extended time stamp extra field format support (Christian)
29. Support for rsxnt-emx 0.9c win32 compiler (Christian)
30. Use izshr017b (Christian)
------------------------ March 11th 1997 version 2.2f ------------------
 1. Move makefile.emx, rsxwinnt.h and zip.def to win32 subdir (Kai Uwe)
 2. Add win32 target to makefile.os2 to allow cross compilation (Kai Uwe)
 3. Fix NTSD_EAS link time failures with win32 (Paul)
 4. Fix buffer freed too early in password verification code (Mike)
 5. Remove unix/zipgrep and man/zipgrep.1 (sanvila@ctv.es)
 6. Only use crc_i386.o when we're using an x86 (Onno, Mark)
 7. Remove carriage returns from amiga/crc_68.a (Paul)
 8. New windll from Mike
 9. Fix typo in os2/os2zip.c (Kai Uwe)
10. Don't use ctime (last file status change) for unix and qdos cross compile
    (Greg)
11. added gccwin32 crosscompilation target (RSXNT) to os2/makefile.os2 (Kai Uwe)
12. fixed the OS/2 file attribute and time stamp generation for zipping
    stdin ("-") (Kai Uwe)
13. fixed the atime and ctime stat fields for the OS/2 Watcom C library
    (Kai Uwe)
14. added atime and ctime support for the UT extra field when generated under
    OS/2, the atime and ctime values are only stored when zipping (Kai Uwe)
15. qdos patches from Jonathan Hudson mainly for extended time flag handling
16. amiga aztec compiler bug workaround (Paul)
17. fix -v output of zipcloak, zipnote and zipsplit (Paul)
18. new amiga/makefile.azt with targets for debug versions (Paul)
------------------------ March 31st 1997 version 2.2g ------------------
 1. remove -I/usr/local/include from unix/Makefile (Chris)
 2. Update versinfolines in revision.h (Greg)
 3. change 1U to 0x1 to accomodate non ANSI compilers (Onno, Rodney Brown)
 4. win32zip.c: cast buffer parameter in memcompress() to char * (Mike)
 5. remove beos/zipgrep (Chris)
 6. correct the -e password verification check in zip.c (Christian)

zipnote.c
/*
  zipnote.c - Zip 3

  Copyright (c) 1990-2008 Info-ZIP.  All rights reserved.

  See the accompanying file LICENSE, version 2007-Mar-4 or later
  (the contents of which are also included in zip.h) for terms of use.
  If, for some reason, all these files are missing, the Info-ZIP license
  also may be found at:  ftp://ftp.info-zip.org/pub/infozip/license.html
*/
/*
 *  zipnote.c by Mark Adler.
 */
#define __ZIPNOTE_C

#ifndef UTIL
#define UTIL
#endif
#include "zip.h"
#define DEFCPYRT        /* main module: enable copyright string defines! */
#include "revision.h"
#include <signal.h>

/* Calculate size of static line buffer used in write (-w) mode. */
#define WRBUFSIZ 2047
/* The line buffer size should be at least as large as FNMAX. */
#if FNMAX > WRBUFSIZ
#  undef WRBUFSIZ
#  define WRBUFSIZ FNMAX
#endif

/* Character to mark zip entry names in the comment file */
#define MARK '@'
#define MARKE " (comment above this line)"
#define MARKZ " (zip file comment below this line)"

/* Temporary zip file pointer */
local FILE *tempzf;


/* Local functions */
local void handler OF((int));
local void license OF((void));
local void help OF((void));
local void version_info OF((void));
local void putclean OF((char *, extent));
/* getline name conflicts with GNU getline() function */
local char *zgetline OF((char *, extent));
local int catalloc OF((char * far *, char *));
int main OF((int, char **));

/* keep compiler happy until implement long options - 11/4/2003 EG */
struct option_struct far options[] = {
  /* short longopt        value_type        negatable        ID    name */
    {"h",  "help",        o_NO_VALUE,       o_NOT_NEGATABLE, 'h',  "help"},
    /* the end of the list */
    {NULL, NULL,          o_NO_VALUE,       o_NOT_NEGATABLE, 0,    NULL} /* end has option_ID = 0 */
  };

#ifdef MACOS
#define ziperr(c, h)    zipnoteerr(c, h)
#define zipwarn(a, b)   zipnotewarn(a, b)

void zipnoteerr(int c, ZCONST char *h);
void zipnotewarn(ZCONST char *a, ZCONST char *b);
#endif

#ifdef QDOS
#define exit(p1) QDOSexit()
#endif

int set_filetype(out_path)
  char *out_path;
{
#ifdef __BEOS__
  /* Set the filetype of the zipfile to "application/zip" */
  setfiletype( out_path, "application/zip" );
#endif

#ifdef __ATHEOS__
  /* Set the filetype of the zipfile to "application/x-zip" */
  setfiletype(out_path, "application/x-zip");
#endif

#ifdef MACOS
  /* Set the Creator/Type of the zipfile to 'IZip' and 'ZIP ' */
  setfiletype(out_path, 'IZip', 'ZIP ');
#endif

#ifdef RISCOS
  /* Set the filetype of the zipfile to &DDC */
  setfiletype(out_path, 0xDDC);
#endif
  return ZE_OK;
}

/* rename a split
 * A split has a tempfile name until it is closed, then
 * here rename it as out_path the final name for the split.
 */
int rename_split(temp_name, out_path)
  char *temp_name;
  char *out_path;
{
  int r;
  /* Replace old zip file with new zip file, leaving only the new one */
  if ((r = replace(out_path, temp_name)) != ZE_OK)
  {
    zipwarn("new zip file left as: ", temp_name);
    free((zvoid *)tempzip);
    tempzip = NULL;
    ZIPERR(r, "was replacing split file");
  }
  if (zip_attributes) {
    setfileattr(out_path, zip_attributes);
  }
  return ZE_OK;
}

void zipmessage_nl(a, nl)
ZCONST char *a;     /* message string to output */
int nl;             /* 1 = add nl to end */
/* If nl false, print a message to mesg without new line.
   If nl true, print and add new line.  If logfile is
   open then also write message to log file. */
{
  if (noisy) {
    fprintf(mesg, "%s", a);
    if (nl) {
      fprintf(mesg, "\n");
      mesg_line_started = 0;
    } else {
      mesg_line_started = 1;
    }
    fflush(mesg);
  }
}

void zipmessage(a, b)
ZCONST char *a, *b;     /* message strings juxtaposed in output */
/* Print a message to mesg and flush.  Also write to log file if
   open.  Write new line first if current line has output already. */
{
  if (noisy) {
    if (mesg_line_started)
      fprintf(mesg, "\n");
    fprintf(mesg, "%s%s\n", a, b);
    mesg_line_started = 0;
    fflush(mesg);
  }
}

void ziperr(c, h)
int c;                  /* error code from the ZE_ class */
ZCONST char *h;         /* message about how it happened */
/* Issue a message for the error, clean up files and memory, and exit. */
{
  if (PERR(c))
    perror("zipnote error");
  fprintf(mesg, "zipnote error: %s (%s)\n", ZIPERRORS(c), h);
  if (tempzf != NULL)
    fclose(tempzf);
  if (tempzip != NULL)
  {
    destroy(tempzip);
    free((zvoid *)tempzip);
  }
  if (zipfile != NULL)
    free((zvoid *)zipfile);
  EXIT(c);
}


local void handler(s)
int s;                  /* signal number (ignored) */
/* Upon getting a user interrupt, abort cleanly using ziperr(). */
{
#ifndef MSDOS
  putc('\n', mesg);
#endif /* !MSDOS */
  ziperr(ZE_ABORT, "aborting");
  s++;                                  /* keep some compilers happy */
}


void zipwarn(a, b)
ZCONST char *a, *b;     /* message strings juxtaposed in output */
/* Print a warning message to mesg (usually stderr) and return. */
{
  fprintf(mesg, "zipnote warning: %s%s\n", a, b);
}


local void license()
/* Print license information to stdout. */
{
  extent i;             /* counter for copyright array */

  for (i = 0; i < sizeof(swlicense)/sizeof(char *); i++)
    puts(swlicense[i]);
}


local void help()
/* Print help (along with license info) to stdout. */
{
  extent i;             /* counter for help array */

  /* help array */
  static ZCONST char *text[] = {
"",
"ZipNote %s (%s)",
#ifdef VM_CMS
"Usage:  zipnote [-w] [-q] [-b fm] zipfile",
#else
"Usage:  zipnote [-w] [-q] [-b path] zipfile",
#endif
"  the default action is to write the comments in zipfile to stdout",
"  -w   write the zipfile comments from stdin",
#ifdef VM_CMS
"  -b   use \"fm\" as the filemode for the temporary zip file",
#else
"  -b   use \"path\" for the temporary zip file",
#endif
"  -q   quieter operation, suppress some informational messages",
"  -h   show this help    -v   show version info    -L   show software license",
"",
"Example:",
#ifdef VMS
"     define/user sys$output foo.tmp",
"     zipnote foo.zip",
"     edit foo.tmp",
"     ... then you edit the comments, save, and exit ...",
"     define/user sys$input foo.tmp",
"     zipnote -w foo.zip",
#else
#ifdef RISCOS
"     zipnote foo/zip > foo/tmp",
"     <!Edit> foo/tmp",
"     ... then you edit the comments, save, and exit ...",
"     zipnote -w foo/zip < foo/tmp",
#else
#ifdef VM_CMS
"     zipnote foo.zip > foo.tmp",
"     xedit foo tmp",
"     ... then you edit the comments, save, and exit ...",
"     zipnote -w foo.zip < foo.tmp",
#else
"     zipnote foo.zip > foo.tmp",
"     ed foo.tmp",
"     ... then you edit the comments, save, and exit ...",
"     zipnote -w foo.zip < foo.tmp",
#endif /* VM_CMS */
#endif /* RISCOS */
#endif /* VMS */
"",
"  \"@ name\" can be followed by an \"@=newname\" line to change the name"
  };

  for (i = 0; i < sizeof(copyright)/sizeof(char *); i++) {
    printf(copyright[i], "zipnote");
    putchar('\n');
  }
  for (i = 0; i < sizeof(text)/sizeof(char *); i++)
  {
    printf(text[i], VERSION, REVDATE);
    putchar('\n');
  }
}

/*
 * XXX put this in version.c
 */

local void version_info()
/* Print verbose info about program version and compile time options
   to stdout. */
{
  extent i;             /* counter in text arrays */

  /* Options info array */
  static ZCONST char *comp_opts[] = {
#ifdef DEBUG
    "DEBUG",
#endif
    NULL
  };

  for (i = 0; i < sizeof(copyright)/sizeof(char *); i++)
  {
    printf(copyright[i], "zipnote");
    putchar('\n');
  }

  for (i = 0; i < sizeof(versinfolines)/sizeof(char *); i++)
  {
    printf(versinfolines[i], "ZipNote", VERSION, REVDATE);
    putchar('\n');
  }

  version_local();

  puts("ZipNote special compilation options:");
  for (i = 0; (int)i < (int)(sizeof(comp_opts)/sizeof(char *) - 1); i++)
  {
    printf("        %s\n",comp_opts[i]);
  }
  if (i == 0)
      puts("        [none]");
}


local void putclean(s, n)
char *s;                /* string to write to stdout */
extent n;               /* length of string */
/* Write the string s to stdout, filtering out control characters that are
   not tab or newline (mainly to remove carriage returns), and prefix MARK's
   and backslashes with a backslash.  Also, terminate with a newline if
   needed. */
{
  int c;                /* next character in string */
  int e;                /* last character written */

  e = '\n';                     /* if empty, write nothing */
  while (n--)
  {
    c = *(uch *)s++;
    if (c == MARK || c == '\\')
      putchar('\\');
    if (c >= ' ' || c == '\t' || c == '\n')
      { e=c; putchar(e); }
  }
  if (e != '\n')
    putchar('\n');
}


local char *zgetline(buf, size)
char *buf;
extent size;
/* Read a line of text from stdin into string buffer 'buf' of size 'size'.
   In case of buffer overflow or EOF, a NULL pointer is returned. */
{
    char *line;
    unsigned len;

    line = fgets(buf, size, stdin);
    if (line != NULL && (len = strlen(line)) > 0) {
        if (len == size-1 && line[len-1] != '\n') {
            /* buffer is full and record delimiter not seen -> overflow */
            line = NULL;
        } else {
            /* delete trailing record delimiter */
            if (line[len-1] == '\n') line[len-1] = '\0';
        }
    }
    return line;
}


local int catalloc(a, s)
char * far *a;          /* pointer to a pointer to a malloc'ed string */
char *s;                /* string to concatenate on a */
/* Concatentate the string s to the malloc'ed string pointed to by a.
   Preprocess s by removing backslash escape characters. */
{
  char *p;              /* temporary pointer */
  char *q;              /* temporary pointer */

  for (p = q = s; *q; *p++ = *q++)
    if (*q == '\\' && *(q+1))
      q++;
  *p = 0;
  if ((p = malloc(strlen(*a) + strlen(s) + 3)) == NULL)
    return ZE_MEM;
  strcat(strcat(strcpy(p, *a), **a ? "\r\n" : ""), s);
  free((zvoid *)*a);
  *a = p;
  return ZE_OK;
}


#ifndef USE_ZIPNOTEMAIN
int main(argc, argv)
#else
int zipnotemain(argc, argv)
#endif
int argc;               /* number of tokens in command line */
char **argv;            /* command line tokens */
/* Write the comments in the zipfile to stdout, or read them from stdin. */
{
  char abf[WRBUFSIZ+1]; /* input line buffer */
  char *a;              /* pointer to line buffer or NULL */
  zoff_t c;             /* start of central directory */
  int k;                /* next argument type */
  char *q;              /* steps through option arguments */
  int r;                /* arg counter, temporary variable */
  zoff_t s;             /* length of central directory */
  int t;                /* attributes of zip file */
  int w;                /* true if updating zip file from stdin */
  FILE *x;              /* input file for testing if can write it */
  struct zlist far *z;  /* steps through zfiles linked list */

#ifdef THEOS
  setlocale(LC_CTYPE, "I");
#endif

#ifdef UNICODE_SUPPORT
# ifdef UNIX
  /* For Unix, set the locale to UTF-8.  Any UTF-8 locale is
     OK and they should all be the same.  This allows seeing,
     writing, and displaying (if the fonts are loaded) all
     characters in UTF-8. */
  {
    char *loc;

    /*
      loc = setlocale(LC_CTYPE, NULL);
      printf("  Initial language locale = '%s'\n", loc);
    */

    loc = setlocale(LC_CTYPE, "en_US.UTF-8");

    /*
      printf("langinfo %s\n", nl_langinfo(CODESET));
    */

    if (loc != NULL) {
      /* using UTF-8 character set so can set UTF-8 GPBF bit 11 */
      using_utf8 = 1;
      /*
        printf("  Locale set to %s\n", loc);
      */
    } else {
      /*
        printf("  Could not set Unicode UTF-8 locale\n");
      */
    }
  }
# endif
#endif

  /* If no args, show help */
  if (argc == 1)
  {
    help();
    EXIT(ZE_OK);
  }

  /* Direct info messages to stderr; stdout is used for data output. */
  mesg = stderr;

  init_upper();           /* build case map table */

  /* Go through args */
  zipfile = tempzip = NULL;
  tempzf = NULL;
  signal(SIGINT, handler);
#ifdef SIGTERM              /* AMIGA has no SIGTERM */
  signal(SIGTERM, handler);
#endif
#ifdef SIGABRT
  signal(SIGABRT, handler);
#endif
#ifdef SIGBREAK
  signal(SIGBREAK, handler);
#endif
#ifdef SIGBUS
  signal(SIGBUS, handler);
#endif
#ifdef SIGILL
  signal(SIGILL, handler);
#endif
#ifdef SIGSEGV
  signal(SIGSEGV, handler);
#endif
  k = w = 0;
  for (r = 1; r < argc; r++)
    if (*argv[r] == '-') {
      if (argv[r][1])
        for (q = argv[r]+1; *q; q++)
          switch (*q)
          {
            case 'b':   /* Specify path for temporary file */
              if (k)
                ziperr(ZE_PARMS, "use -b before zip file name");
              else
                k = 1;          /* Next non-option is path */
              break;
            case 'h':   /* Show help */
              help();  EXIT(ZE_OK);
            case 'l':  case 'L':  /* Show copyright and disclaimer */
              license();  EXIT(ZE_OK);
            case 'q':   /* Quiet operation, suppress info messages */
              noisy = 0;  break;
            case 'v':   /* Show version info */
              version_info();  EXIT(ZE_OK);
            case 'w':
              w = 1;  break;
            default:
              ziperr(ZE_PARMS, "unknown option");
          }
      else
        ziperr(ZE_PARMS, "zip file cannot be stdin");
    } else
      if (k == 0)
      {
        if (zipfile == NULL)
        {
          if ((zipfile = ziptyp(argv[r])) == NULL)
            ziperr(ZE_MEM, "was processing arguments");
        }
        else
          ziperr(ZE_PARMS, "can only specify one zip file");
      }
      else
      {
        tempath = argv[r];
        k = 0;
      }
  if (zipfile == NULL)
    ziperr(ZE_PARMS, "need to specify zip file");

  if ((in_path = malloc(strlen(zipfile) + 1)) == NULL) {
    ziperr(ZE_MEM, "input");
  }
  strcpy(in_path, zipfile);

  /* Read zip file */
  if ((r = readzipfile()) != ZE_OK)
    ziperr(r, zipfile);
  if (zfiles == NULL)
    ziperr(ZE_NAME, zipfile);

  /* Put comments to stdout, if not -w */
  if (!w)
  {
    for (z = zfiles; z != NULL; z = z->nxt)
    {
      printf("%c %s\n", MARK, z->zname);
      putclean(z->comment, z->com);
      printf("%c%s\n", MARK, MARKE);
    }
    printf("%c%s\n", MARK, MARKZ);
    putclean(zcomment, zcomlen);
    EXIT(ZE_OK);
  }

  /* If updating comments, make sure zip file is writeable */
  if ((x = fopen(zipfile, "a")) == NULL)
    ziperr(ZE_CREAT, zipfile);
  fclose(x);
  t = getfileattr(zipfile);

  /* Process stdin, replacing comments */
  z = zfiles;
  while ((a = zgetline(abf, WRBUFSIZ+1)) != NULL &&
         (a[0] != MARK || strcmp(a + 1, MARKZ)))
  {                                     /* while input and not file comment */
    if (a[0] != MARK || a[1] != ' ')    /* better be "@ name" */
      ziperr(ZE_NOTE, "unexpected input");
    while (z != NULL && strcmp(a + 2, z->zname))
      z = z->nxt;                       /* allow missing entries in order */
    if (z == NULL)
      ziperr(ZE_NOTE, "unknown entry name");
    if ((a = zgetline(abf, WRBUFSIZ+1)) != NULL && a[0] == MARK && a[1] == '=')
    {
      if (z->name != z->iname)
        free((zvoid *)z->iname);
      if ((z->iname = malloc(strlen(a+1))) == NULL)
        ziperr(ZE_MEM, "was changing name");
#ifdef EBCDIC
      strtoasc(z->iname, a+2);
#else
      strcpy(z->iname, a+2);
#endif

/*
 * Don't update z->nam here, we need the old value a little later.....
 * The update is handled in zipcopy().
 */
      a = zgetline(abf, WRBUFSIZ+1);
    }
    if (z->com)                         /* change zip entry comment */
      free((zvoid *)z->comment);
    z->comment = malloc(1);  *(z->comment) = 0;
    while (a != NULL && *a != MARK)
    {
      if ((r = catalloc(&(z->comment), a)) != ZE_OK)
        ziperr(r, "was building new zipentry comments");
      a = zgetline(abf, WRBUFSIZ+1);
    }
    z->com = strlen(z->comment);
    z = z->nxt;                         /* point to next entry */
  }
  if (a != NULL)                        /* change zip file comment */
  {
    zcomment = malloc(1);  *zcomment = 0;
    while ((a = zgetline(abf, WRBUFSIZ+1)) != NULL)
      if ((r = catalloc(&zcomment, a)) != ZE_OK)
        ziperr(r, "was building new zipfile comment");
    zcomlen = strlen(zcomment);
  }

  /* Open output zip file for writing */
#if defined(UNIX) && !defined(NO_MKSTEMP)
  {
    int yd;
    int i;

    /* use mkstemp to avoid race condition and compiler warning */

    if (tempath != NULL)
    {
      /* if -b used to set temp file dir use that for split temp */
      if ((tempzip = malloc(strlen(tempath) + 12)) == NULL) {
        ZIPERR(ZE_MEM, "allocating temp filename");
      }
      strcpy(tempzip, tempath);
      if (lastchar(tempzip) != '/')
        strcat(tempzip, "/");
    }
    else
    {
      /* create path by stripping name and appending template */
      if ((tempzip = malloc(strlen(zipfile) + 12)) == NULL) {
      ZIPERR(ZE_MEM, "allocating temp filename");
      }
      strcpy(tempzip, zipfile);
      for(i = strlen(tempzip); i > 0; i--) {
        if (tempzip[i - 1] == '/')
          break;
      }
      tempzip[i] = '\0';
    }
    strcat(tempzip, "ziXXXXXX");

    if ((yd = mkstemp(tempzip)) == EOF) {
      ZIPERR(ZE_TEMP, tempzip);
    }
    if ((tempzf = y = fdopen(yd, FOPW)) == NULL) {
      ZIPERR(ZE_TEMP, tempzip);
    }
  }
#else
  if ((tempzf = y = fopen(tempzip = tempname(zipfile), FOPW)) == NULL)
    ziperr(ZE_TEMP, tempzip);
#endif

  /* Open input zip file again, copy preamble if any */
  if ((in_file = fopen(zipfile, FOPR)) == NULL)
    ziperr(ZE_NAME, zipfile);

  if (zipbeg && (r = bfcopy(zipbeg)) != ZE_OK)
    ziperr(r, r == ZE_TEMP ? tempzip : zipfile);
  tempzn = zipbeg;

  /* Go through local entries, copying them over as is */
  fix = 3; /* needed for zipcopy if name changed */
  for (z = zfiles; z != NULL; z = z->nxt) {
    if ((r = zipcopy(z)) != ZE_OK)
      ziperr(r, "was copying an entry");
  }
  fclose(x);

  /* Write central directory and end of central directory with new comments */
  if ((c = zftello(y)) == (zoff_t)-1)    /* get start of central */
    ziperr(ZE_TEMP, tempzip);
  for (z = zfiles; z != NULL; z = z->nxt)
    if ((r = putcentral(z)) != ZE_OK)
      ziperr(r, tempzip);
  if ((s = zftello(y)) == (zoff_t)-1)    /* get end of central */
    ziperr(ZE_TEMP, tempzip);
  s -= c;                       /* compute length of central */
  if ((r = putend((zoff_t)zcount, s, c, zcomlen, zcomment)) != ZE_OK)
    ziperr(r, tempzip);
  tempzf = NULL;
  if (fclose(y))
    ziperr(ZE_TEMP, tempzip);
  if ((r = replace(zipfile, tempzip)) != ZE_OK)
  {
    zipwarn("new zip file left as: ", tempzip);
    free((zvoid *)tempzip);
    tempzip = NULL;
    ziperr(r, "was replacing the original zip file");
  }
  free((zvoid *)tempzip);
  tempzip = NULL;
  setfileattr(zipfile, t);
#ifdef RISCOS
  /* Set the filetype of the zipfile to &DDC */
  setfiletype(zipfile,0xDDC);
#endif
  free((zvoid *)zipfile);
  zipfile = NULL;

  /* Done! */
  RETURN(0);
}
